course_flow/decorators.py

```diff
@@ -104,65 +75,143 @@ def _wrapped_view(request, model=model, *args, **kwargs):
     return wrapped_view
 
 
-def is_owner_or_none(model):
+def test_object_permission(instance, user, permission):
+    if hasattr(instance, "get_subclass"):
+        instance = instance.get_subclass()
+    if instance.author == user:
+        return True
+    if permission == ObjectPermission.PERMISSION_VIEW:
+        if instance.published == True:
+            return True
+        if (
+            ObjectPermission.objects.filter(
+                user=user,
+                object_id=instance.id,
+                content_type=ContentType.objects.get_for_model(instance),
+                permission_type=ObjectPermission.PERMISSION_EDIT,
+            ).count()
+            > 0
+        ):
+            return True
+    return (
+        ObjectPermission.objects.filter(
+            user=user,
+            object_id=instance.id,
+            content_type=ContentType.objects.get_for_model(instance),
+            permission_type=permission,
+        ).count()
+        > 0
+    )
+
+
+def test_objects_permission(instances, user, permission):
+    object_permissions = [
+        test_object_permission(x, user, permission) for x in instances
+    ]
+    return reduce(lambda a, b: a | b, object_permissions)
+
+
+def test_special_case_delete_permission(model_data, user):
+    instance = get_model_from_str(model_data["model"]).objects.get(
+        id=model_data["id"]
+    )
+    #    if (
+    #        model_data["model"] == "outcome"
+    #        and OutcomeWorkflow.objects.filter(outcome=instance).count() == 0
+    #    ):
+    #        permission_objects = instance.get_permission_objects()
+    #        return test_objects_permission(
+    #            permission_objects, user, ObjectPermission.PERMISSION_EDIT
+    #        )
+    if model_data["model"] == "project":
+        return instance.author == user
+    else:
+        if hasattr(instance, "get_subclass"):
+            instance = instance.get_subclass()
+        if instance.get_project() is None:
```

Comment:
Again I don't see how we could. Are you talking about specifically the get_subclass() part? That's to get the workflow's subclass, which is what holds the author. The rest deals with specific cases of workflows that are not in projects (which can exist if they are a template/strategy) and those that are (where we should be able to delete them if we are their owner or their project's owner).

Reply:
I was thinking along same lines as above: the default return contains same logic as the if clause.  If `instance.get_project() is None return instance.author == user` and if it isn't `None` still return `instance.author==user`.  Maybe the `or` complicates things.

Reply:
The or does indeed complicate things. Because if it does have a project, we want to return true if EITHER its own author or the project author match. I guess we could reverse this with:
`if instance.get_project() is not None and instance.get_project().author==user:
    return True
else:
    return instance.author==user`
but that doesn't really simplify anything.